src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -225,6 +225,89 @@ double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Evaluate(
   return res;
 }
 
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Evaluate(
+    const arma::mat& parameters, const size_t begin, const size_t batchSize)
+{
+  return Evaluate(parameters, begin, batchSize, true);
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+template<typename GradType>
+double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::
+EvaluateWithGradient(const arma::mat& parameters, GradType& gradient)
+{
+  double res = 0;
+  for (size_t i = 0; i < predictors.n_cols; ++i)
+    res += EvaluateWithGradient(parameters, i, gradient, 1, true);
+
+  return res;
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+template<typename GradType>
+double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::
+EvaluateWithGradient(const arma::mat& /* parameters */,
+                     const size_t begin,
+                     GradType& gradient,
+                     const size_t batchSize,
+                     const bool deterministic)
+{
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+      ResetParameters();
+
+    gradient = arma::zeros<arma::mat>(parameter.n_rows, parameter.n_cols);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  if (deterministic != this->deterministic)
+  {
+    this->deterministic = deterministic;
+    ResetDeterministic();
+  }
+
+  Forward(std::move(predictors.cols(begin, begin + batchSize - 1)));
+  double res = outputLayer.Forward(
+      std::move(boost::apply_visitor(outputParameterVisitor, network.back())),
+      std::move(responses.cols(begin, begin + batchSize - 1)));
+
+  for (size_t i = 0; i < network.size(); ++i)
+  {
+    res += boost::apply_visitor(lossVisitor, network[i]);
+  }
+
+  outputLayer.Backward(
+      std::move(boost::apply_visitor(outputParameterVisitor, network.back())),
+      std::move(responses.cols(begin, begin + batchSize - 1)),
+      std::move(error));
+
+  Backward();
+  ResetGradients(gradient);
+  Gradient(std::move(predictors.cols(begin, begin + batchSize - 1)));
```

Comment:
https://github.com/mlpack/mlpack/blob/4c19146cb7b0566226fb16dc49051af664c74b79/src/mlpack/methods/ann/ffn_impl.hpp#L248

This call in the ``Gradient(..)`` method can be avoided, since it's what we already did in line 277-285. So, I guess we can just implement our own Gradient function here, without the extra ``Evaluate``call, let me know what you think.